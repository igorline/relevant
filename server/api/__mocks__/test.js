// exports.test = 'test';

module.exports = 'test';
