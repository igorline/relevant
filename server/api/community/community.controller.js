import Community from './community.model';
import CommunityMember from './community.member.model';
import User from '../user/user.model';

const RESERVED = [
  'user',
  'admin',
  'info',
  'api',
  'img',
  'fonts',
  'files',
  'home',
  'undefined'
];

export async function findOne(req, res, next) {
  try {
    const { user } = req;
    const { slug } = req.params;
    const community = await Community.findOne({ slug, inactive: { $ne: true } });
    if (!community) throw new Error(`Community ${community} doesn't exist`);

    if (community.private) {
      if (!user) throw new Error('This community is private');
      const member = await CommunityMember.findOne({
        communityId: community._id,
        user: user._id
      });
      if (!member) throw new Error('This community is private');
    }

    res.status(200).json(community);
  } catch (err) {
    next(err);
  }
}

// uses middleware for server-sider rendering
export async function index(req) {
  const { user } = req;
  const { community } = req.query;
  const onlyPublic = user && user.role === 'admin' ? {} : { private: { $ne: true } };
  const onlyVisible = user && user.role === 'admin' ? {} : { hidden: { $ne: true } };

  const communties = await Community.find({
    inactive: { $ne: true },
    ...onlyPublic,
    $or: [onlyVisible, { slug: community }]
  }).populate({
    path: 'admins',
    match: { role: 'admin' }
  });

  // find private communities where user is a member
  let privateCommunities = [];
  if (user) {
    const memberships = await CommunityMember.find({ user: user._id }).populate({
      path: 'communityId',
      match: { inactive: { $ne: true } }
    });
    privateCommunities = memberships
    .filter(m => m.communityId)
    .filter(
      m =>
        m.communityId.private === true ||
          (m.communityId.hidden === true && m.communityId.slug !== community)
    )
    .map(m => m.communityId);
  }
  return [...communties, ...privateCommunities].map(c => c.toObject());
}

export async function members(req, res, next) {
  try {
    const { user } = req;
    let blocked = [];
    if (user) {
      blocked = [...user.blocked, ...user.blockedBy];
    }
    // const userId = user ? user._id : null;
    const limit = req.query.limit ? parseInt(req.query.limit, 10) : 20;
    const skip = req.query.skip ? parseInt(req.query.skip, 10) : 0;
    const community = req.params.slug;
    // const isMember = await CommunityMember.findOne({ community, user: userId });
    const users = await CommunityMember.find({
      community,
      'user.embeddedUser._id': {
        $nin: blocked
      }
    })
    .sort({ role: 1, reputation: -1 })
    .limit(limit)
    .skip(skip);
    res.status(200).json(users || []);
  } catch (err) {
    next(err);
  }
}

export async function memberSearch(req, res, next) {
  try {
    let blocked = [];
    const { user } = req;
    if (user) {
      blocked = [...user.blocked, ...user.blockedBy];
    }

    const { search, limit } = req.query;
    const name = new RegExp(search, 'i');
    const query = {
      $and: [
        { $or: [{ 'embeddedUser.name': name }, { 'embeddedUser.handle': name }] },
        { 'embeddedUser._id': { $nin: blocked } }
      ]
    };
    const community = req.params.slug;
    CommunityMember.find({ community, ...query })
    .sort({ role: 1, reputation: -1 })
    .limit(parseInt(limit, 10))
    .then(users => {
      res.status(200).json(users || []);
    });
    // res.status(200).json(users);
    // .catch(next);
  } catch (err) {
    next(err);
  }
}

export async function membership(req, res, next) {
  try {
    const user = req.user._id;
    const m = await CommunityMember.find({ user }).sort('role reputation');
    res.status(200).json(m);
  } catch (err) {
    next(err);
  }
}

export async function join(req, res, next) {
  try {
    const userId = req.user._id;
    const { slug } = req.params;
    const community = await Community.findOne({ slug });
    const member = await community.join(userId);
    res.status(200).json(member);
  } catch (err) {
    next(err);
  }
}

export async function leave(req, res, next) {
  try {
    const userId = req.user._id;
    const { slug } = req.params;
    const community = await Community.findOne({ slug });
    await community.leave(userId);
    res.sendStatus(200);
  } catch (err) {
    next(err);
  }
}

export async function showAdmins(req, res, next) {
  try {
    const { slug } = req.params;
    const admins = await CommunityMember.find({ slug, role: 'admin' });
    res.status(200).json(admins);
  } catch (err) {
    next(err);
  }
}

export async function create(req, res, next) {
  try {
    // for no only admins create communities
    // TODO relax this and community creator as admin
    const { user } = req;
    if (!req.user || !req.user.role === 'admin') {
      throw new Error("You don't have permission to create a community");
    }
    let community = req.body;
    let { admins } = community;
    admins = await User.find({ handle: { $in: admins } }, '_id');
    community.slug = community.slug.toLowerCase();
    if (RESERVED.indexOf(community.slug) > -1) throw new Error('Reserved slug');
    if (!admins || !admins.length) throw new Error('Please set community admins');
    community = new Community(community);
    community = await community.save();

    if (user.role !== 'admin') await community.join(user._id, 'superAdmin');

    // TODO - this should create an invitation!
    admins = admins.map(async admin => {
      try {
        return await community.join(admin._id, 'admin');
      } catch (err) {
        throw err;
      }
    });

    admins = await Promise.all(admins);

    res.status(200).json(community);
  } catch (err) {
    next(err);
  }
}

export async function update(req, res, next) {
  try {
    // for no only admins create communities
    const updatedCommunity = req.body;
    const { admins, superAdmins } = updatedCommunity;
    const { user } = req;

    if (!updatedCommunity || !user) return;
    const member = await CommunityMember.findOne({
      communityId: updatedCommunity._id,
      user: user._id
    });

    const currentAdmins = await CommunityMember.find({
      communityId: updatedCommunity._id,
      role: 'admin'
    });

    const currentAdminsList = currentAdmins.map(a => a.embeddedUser.handle);
    let newAdmins = admins.filter(a => !currentAdminsList.includes(a));

    const canEdit = user.role === 'admin' || (member && member.superAdmin);

    if (!canEdit) {
      throw new Error("You don't have permission to edit a community");
    }

    newAdmins = await User.find({ handle: { $in: newAdmins } }, '_id');

    let community = await Community.findOne({ _id: updatedCommunity._id });
    community.set({
      image: updatedCommunity.image,
      name: updatedCommunity.name,
      topics: updatedCommunity.topics,
      description: updatedCommunity.description,
      channels: updatedCommunity.channels,
      private: updatedCommunity.private,
      hidden: updatedCommunity.hidden
    });
    community = await community.save();

    // TODO - this should create an invitation!
    newAdmins = newAdmins.map(async admin => {
      try {
        const role = superAdmins.includes(admin._id) ? 'superAdmin' : 'admin';
        return await community.join(admin._id, role);
      } catch (err) {
        throw err;
      }
    });

    await Promise.all(newAdmins);

    let removeAdmins;
    if (user.role === 'admin') {
      removeAdmins = currentAdminsList.filter(a => !admins.includes(a));
      removeAdmins = await User.find({ handle: { $in: removeAdmins } }, '_id');
      removeAdmins = removeAdmins.map(u => u._id.toString());
      await CommunityMember.updateMany(
        { user: { $in: removeAdmins } },
        { role: 'user', superAdmin: false },
        { multi: true }
      );
    }

    await CommunityMember.updateMany(
      { 'embeddedUser.handle': { $in: admins } },
      { superAdmin: false },
      { multi: true }
    );

    await CommunityMember.updateMany(
      { 'embeddedUser.handle': { $in: superAdmins }, role: 'admin' },
      { superAdmin: true },
      { multi: true }
    );

    community.admins = await CommunityMember.find({
      communityId: community._id,
      role: 'admin'
    });

    res.status(200).json(community);
  } catch (err) {
    next(err);
  }
}

export async function remove(req, res, next) {
  try {
    if (process.env.NODE_ENV === 'production') {
      throw new Error('deleting communities is disabled in production');
    }
    const { user } = req;
    if (user.role !== 'admin') {
      throw new Error("you don't have permission to delete this community");
    }

    const userId = req.user._id;
    const { slug } = req.params;
    // check that user is an admin
    const admin = CommunityMember.findOne({
      community: slug,
      user: userId,
      role: 'admin'
    });

    if (!admin) throw new Error('you need to be a community admin to do this');

    // await Community.findOne({ slug }).remove().exec();
    await Community.findOneAndUpdate({ slug }, { inactive: true }, { new: true });

    res.status(200).json('removed');
  } catch (err) {
    next(err);
  }
}
