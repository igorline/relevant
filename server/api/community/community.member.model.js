const mongoose = require('mongoose');

const { Schema } = mongoose;

const CommunityMemberSchema = new Schema(
  {
    user: { type: Schema.Types.ObjectId, ref: 'User', index: true },
    embeddedUser: {
      _id: String,
      name: String,
      image: String,
      handle: String
    },
    communityId: { type: Schema.Types.ObjectId, ref: 'Community' },
    community: String,
    superAdmin: { type: Boolean, default: false },
    role: { type: String, default: 'user' },
    reputation: { type: Number, default: 0 },
    invites: { type: Number, default: 0 },
    degree: { type: Number, default: 0 },
    postDegree: { type: Number, default: 0 },
    pagerank: { type: Number, default: 0 },
    pagerankNeg: { type: Number, default: 0 },
    pagerankRaw: { type: Number, default: 0 },
    pagerankRawNeg: { type: Number, default: 0 },
    unread: { type: Number, default: 0 },
    deletedCommunity: { type: Boolean, default: false },
    customAdminWeight: { type: Number },
    defaultWeight: { type: Number },

    totalUsers: Number,
    level: Number,
    percentRank: Number,
    relevanceRecord: [
      {
        relevance: Number,
        time: Date
      }
    ]
  },
  {
    timestamps: true
  }
);

CommunityMemberSchema.index({ community: 1 });
CommunityMemberSchema.index({ communityId: 1 });
CommunityMemberSchema.index({ deletedCommunity: 1 });

CommunityMemberSchema.index({ communityId: 1, deletedCommunity: 1 });
CommunityMemberSchema.index({ communityId: 1, user: 1 });
CommunityMemberSchema.index({ community: 1, user: 1 });

CommunityMemberSchema.index({ community: 1, reputation: -1 });
CommunityMemberSchema.index({ community: 1, reputation: -1, role: 1 });
CommunityMemberSchema.index({ community: 1, reputation: -1, user: 1 });

CommunityMemberSchema.index({ deletedCommunity: 1, user: 1 });

// TODO rep key to search by
CommunityMemberSchema.virtual('repKey').get(function getProfile() {
  return this.user + '_' + this.communityId;
});

// update user relevance and save record
CommunityMemberSchema.methods.updateRelevanceRecord = function updateRelevanceRecord() {
  let { relevanceRecord } = this;
  if (!relevanceRecord) relevanceRecord = [];
  relevanceRecord.unshift({
    time: new Date(),
    relevance: this.pagerank
  });
  relevanceRecord = this.relevanceRecord.slice(0, 10);
  this.relevanceRecord = relevanceRecord;
  return this;
};

export default mongoose.model('CommunityMember', CommunityMemberSchema);
