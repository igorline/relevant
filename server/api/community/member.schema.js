import { composeWithMongoose } from 'graphql-compose-mongoose';
import { withFilter } from 'graphql-subscriptions';
import { pubsub } from 'server/graphql/pubsub';
import { schemaComposer } from 'graphql-compose';
import CommunityMember from './community.member.model';
// import Community from './community.model';

export const UPDATE_UNREAD = 'UPDATE_UNREAD';
const TC = composeWithMongoose(CommunityMember, {});

// async function testSub() {
//   const c = await Community.findOne({ slug: 'relevant' });
//   console.log(c);
//   setInterval(() => {
//     pubsub.publish(UPDATE_UNREAD, { communityId: c._id.toString(), community: c.slug });
//   }, 1000);
// }
// testSub();

// STEP 3: Add needed CRUD User operations to the GraphQL Schema
// via graphql-compose it will be much much easier, with less typing
export const memberQuery = {
  memberOne: TC.getResolver('findOne'),
  memberById: TC.getResolver('findById').wrapResolve(next => rp => {
    return next(rp);
  }),
  // userByIds: TC.getResolver('findByIds'),
  memberMany: TC.getResolver('findMany'),
  // userCount: TC.getResolver('count'),
  // userConnection: TC.getResolver('connection'),
  // userPagination: TC.getResolver('pagination')

  myMemberships: TC.getResolver('findMany').wrapResolve(next => rp => {
    if (!rp.context.user) return null;
    rp.args.filter = {
      ...rp.args.filter,
      user: rp.context.user._id,
      deletedCommunity: { $ne: true }
    };
    return next(rp);
  })
};

function onlyOwner(resolvers) {
  Object.keys(resolvers).forEach(k => {
    resolvers[k] = resolvers[k].wrapResolve(next => async rp => {
      rp.beforeRecordMutate = async function bfRM(doc, rpHook) {
        if (!rpHook.context.user || !doc.user.equals(rpHook.context.user._id)) {
          throw new Error('Forbidden!');
        }
        return doc;
      };
      return next(rp);
    });
  });
  return resolvers;
}

export const memberSubscription = {
  updateUnread: {
    type: TC.getType(),
    args: {
      communities: '[ID!]!'
    },
    description: 'Subscribe to unread posts notifications',
    resolve: payload => payload,
    subscribe: withFilter(
      () => pubsub.asyncIterator(UPDATE_UNREAD),
      (payload, args) => args.communities.includes(payload.communityId)
    )
  }
};

export const memberMutation = {
  // userCreateOne: UserTC.getResolver('createOne'),
  // userCreateMany: UserTC.getResolver('createMany'),
  ...onlyOwner({
    memberUpdateById: TC.getResolver('updateById'),
    memberUpdateOne: TC.getResolver('updateOne'),
    updateUnread: schemaComposer.createResolver({
      name: 'updateUnread',
      type: TC.getResolver('updateById').type,
      kind: TC.getResolver('updateById').kind,
      args: {
        record: `input MemberUnreadInput {
        _id: ID!
        unread: Int!
      }`
      },
      resolve: TC.getResolver('updateById').resolve
    })
  })

  //   userUpdateMany: UserTC.getResolver('updateMany'),
  //   userRemoveById: UserTC.getResolver('removeById'),
  //   userRemoveOne: UserTC.getResolver('removeOne'),
  //   userRemoveMany: UserTC.getResolver('removeMany'),
};
