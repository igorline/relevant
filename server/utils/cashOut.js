import Earnings from '../api/earnings/earnings.model';

export async function logCashOut(user, cashOutAmt, next) {
  try {
    const earnings = await Earnings.find({
      user: user._id,
      status: 'paidout',
      earned: { $gt: 0 }
    });
    const totalPreviousPaidout = earnings.reduce(getTotalEarnings, 0);
    const cashOutLog = new Earnings({
      user: user._id,
      cashOutAttempt: true,
      cashOutAmt,
      prevBalance: user.balance,
      endBalance: user.balance - cashOutAmt,
      totalPreviousPaidout,
      legacyAirdrop: user.legacyAirdrop,
      referralTokens: user.referralTokens,
      airdropTokens: user.airdropTokens,
      status: 'in progress',
      source: 'wallet'
    });
    await cashOutLog.save();
    return cashOutLog;
  } catch (err) {
    return next(new Error('Error logging cashOut', err));
  }
}

export function getTotalEarnings(total, currentVal) {
  total += currentVal.earned;
  return total;
}
