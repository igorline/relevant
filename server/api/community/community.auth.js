import request from 'request';
import { promisify } from 'util';
import { utils } from 'ethers';
import { getTokenBalance } from 'server/utils/contract';
import computePageRank from 'server/pagerank/pagerankCompute';
import Community from './community.model';

const requestAsync = promisify(request);

const getAssetsUrl = address =>
  `https://map-api-direct.foam.space:443/user/${address}/assets`;

const foamToken = '0x4946fcea7c692606e8908002e55a582af44ac121';

// const dummyAddress = '0x222861f16354020F62bBfa0A878B2F047a385576';

const foamParams = {
  auth: {
    tokens: 100,
    points: 5
  }
};

async function initFoamParams() {
  const foam = await Community.findOne({ slug: 'foam' });
  if (!foam || foam.customParams) return;
  await foam.setCustomParams(foamParams);
}

export async function checkAuthRoute(req, res, next) {
  try {
    const { user, communityMember } = req;
    const { communityId } = communityMember;
    await checkCommunityAuth({ user, communityId, communityMember });
    res.status(200).json({ OK: true });
  } catch (err) {
    next(err);
  }
}

export async function checkCommunityAuth({ user, communityId, communityMember }) {
  let community = await Community.findOne({ _id: communityId });
  if (community.slug !== 'foam') return true;
  if (!community.customParams) {
    community = await initFoamParams();
  }
  const { tokens, points } = community.customParams.auth;

  const { ethLogin } = user;
  // FOR TESTING ONLY
  // const ethLogin = dummyAddress;

  if (!ethLogin) {
    throw new Error(
      'You need to connect the Ethereum address you use with FOAM in order to post in this forum.\nYou can connect your address with Metamask via Wallet -> Connect Wallet.'
    );
  }
  const res = await requestAsync({ url: getAssetsUrl(ethLogin) });
  const { verifiedPOIs, pendingPOIs } = JSON.parse(res.body);
  const totalPoints = verifiedPOIs + pendingPOIs || 0;

  if (!totalPoints || totalPoints < points) {
    throw new Error(
      `You can only post in this forum only after you have:\n - added ${points} points of interest to the FOAM map\n - have a balance of at least ${tokens} FOAM tokens`
    );
  }
  const balanceWei = await getTokenBalance({
    address: ethLogin,
    tokenAddress: foamToken
  });

  const balance = utils.formatEther(balanceWei);
  if (balance < tokens)
    throw new Error(
      `You can only post in this forum only after you have:\n - added ${points} points of interest to the FOAM map\n - have a balance of at least ${tokens} FOAM tokens`
    );

  // if (!communityMember || !communityMember.role !== 'admin') {
  // await community.join(user._id, 'admin');
  if (!communityMember) {
    await community.join(user._id);
  }

  if (communityMember.pagerank === 0) {
    computePageRank({
      communityId: community._id,
      community: community.slug
    });
  }

  if (communityMember.defaultWeight !== totalPoints + Math.log(tokens + 1)) {
    communityMember.defaultWeight = totalPoints + Math.log(tokens + 1);
    // communityMember.customAdminWeight = totalPoints + Math.log(tokens + 1);
    await communityMember.save();
  }
  return true;
}
