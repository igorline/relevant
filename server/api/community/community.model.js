import mongoose from 'mongoose';
import { NAME_PATTERN } from 'app/utils/text';
import socketEvent from 'server/socket/socketEvent';

const { Schema } = mongoose;

const CommunitySchema = new Schema(
  {
    name: String,
    slug: { type: String, unique: true, required: true },
    description: String,
    channels: [{ type: String }],
    topics: [{ type: String, ref: 'Tags' }],
    image: String,
    currentPosts: { type: Number, default: 0 },

    rewardFund: { type: Number, default: 0 },
    currentShares: { type: Number, default: 0 },
    topPostShares: { type: Number, default: 0 },

    postCount: { type: Number, default: 0 },
    lastRewardFundUpdate: { type: Date, default: new Date() },

    // TODO - start using this for twitter?
    avgTwitterScore: { type: Number, default: 0 },
    twitterCount: { type: Number, default: 0 },
    lastTwitterUpdate: { type: Date },
    maxUserRank: { type: Number },
    maxPostRank: { type: Number },
    danglingConsumer: { type: Number, default: 0 },
    negConsumer: { type: Number, default: 0 },
    numberOfElements: { type: Number },
    memberCount: { type: Number },
    inactive: Boolean,
    private: { type: Boolean, default: false },
    hidden: { type: Boolean, default: false },
    betEnabled: { type: Boolean, default: true },
    defaultPost: { type: String, default: 'link' },

    customParams: {}
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
  }
);

CommunitySchema.virtual('admins', {
  ref: 'CommunityMember',
  localField: 'slug',
  foreignField: 'community'
});

CommunitySchema.virtual('superAdmins', {
  ref: 'CommunityMember',
  localField: 'slug',
  foreignField: 'community'
});

CommunitySchema.virtual('members', {
  ref: 'CommunityMember',
  localField: 'slug',
  foreignField: 'community'
});

CommunitySchema.index({ slug: 1 });

// Validate _id
CommunitySchema.path('slug').validate(
  slug => NAME_PATTERN.test(slug),
  'Username can only contain letters, numbers, dashes and underscores'
);

CommunitySchema.pre('remove', async function remove(next) {
  const members = await this.model('CommunityMember').find({ community: this.slug });
  await this.model('CommunityMember')
    .deleteMany({ community: this.slug })
    .exec();
  // THIS IS TRICKY BECAUSE OF LEAVE RACE CONDITIONS
  const leave = members.map(async m => this.leave(m.user));
  if (leave) await Promise.all(leave);
  next();
});

CommunitySchema.methods.updateMemeberCount = async function updateMemeberCount() {
  this.memberCount = await this.model('CommunityMember').countDocuments({
    communityId: this._id
  });
  return this.save();
};

CommunitySchema.methods.leave = async function leave(userId) {
  await this.model('CommunityMember').deleteOne({ user: userId, communityId: this._id });
  await this.updateMemeberCount();
};

CommunitySchema.methods.join = async function join(userId, role, dontUpdateCount) {
  const superAdmin = role === 'superAdmin';

  const { _id: communityId, slug: community } = this;
  const user = await this.model('User').findOne({ _id: userId }, 'name image handle');
  if (!user) throw new Error('missing user');

  let member = await this.model('CommunityMember').findOne({
    user: user._id,
    communityId
  });

  if (member && role === 'admin') {
    member.role = 'admin';
    return member.save();
  }

  if (member && superAdmin) {
    member.superAdmin = true;
    return member.save();
  }

  if (member) return member;

  member = {
    user: userId,
    embeddedUser: user,
    communityId,
    community,
    reputation: 0,
    superAdmin,
    role: superAdmin ? 'admin' : role || 'user'
  };

  member = new (this.model('CommunityMember'))(member);
  await member.save();

  if (!dontUpdateCount) {
    await this.updateMemeberCount();
  }

  const memberEvent = {
    _id: user._id,
    type: 'ADD_USER_MEMBERSHIP',
    payload: member
  };
  socketEvent.emit('socketEvent', memberEvent);

  return member;
};

CommunitySchema.methods.setCustomParams = async function setCustomParams(params) {
  this.customParams = params;
  return this.save();
};

export default mongoose.model('Community', CommunitySchema);
