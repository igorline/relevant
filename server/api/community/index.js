import { checkAuthRoute } from './community.auth';

const express = require('express');
const controller = require('./community.controller');
const auth = require('../../auth/auth.service');
const { asyncMiddleware } = require('../../utils/middlewares');

const router = express.Router();

router.get('/:slug/members', auth.blocked(), controller.members);
// Search by embedded user handle and name
router.get('/:slug/members/search', auth.blocked(), controller.memberSearch);
router.get('/membership/:user', auth.isAuthenticated(), controller.membership);
router.get('/', auth.currentUser(), asyncMiddleware(controller.index));
router.get('/:slug', auth.isAuthenticated(), controller.findOne);
router.get(
  '/:slug/checkAuth',
  auth.isAuthenticated(),
  auth.communityMember(),
  checkAuthRoute
);

router.post('/', auth.isAuthenticated(), controller.create);

router.put(
  '/:slug',
  auth.isAuthenticated(),
  auth.communityMember('superAdmin'),
  controller.update
);
router.put('/:slug/join', auth.isAuthenticated(), controller.join);
router.put('/:slug/leave', auth.isAuthenticated(), controller.leave);

router.delete('/:id', auth.isAuthenticated(), controller.remove);

module.exports = router;
