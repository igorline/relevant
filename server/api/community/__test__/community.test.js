import CommunityMember from 'server/api/community/community.member.model';
// import Community from 'server/api/community/community.model';
import { community } from 'app/mockdata';
import { getUsers } from 'server/test/seedData';
import { create, join, leave, update } from 'server/api/community/community.controller';
import { response } from 'jest-mock-express';
import { sanitize } from 'server/test/utils';

const { brandNew } = community;
process.env.TEST_SUITE = 'communities';

describe('Communities', () => {
  let res;
  let req;
  let bob;
  let carol;
  let newCommunityId;
  const next = jest.fn(console.log); // eslint-disable-line

  beforeEach(() => {
    res = response();
    ({ bob, carol } = getUsers());
    req = {
      user: bob,
      body: brandNew
    };
  });

  describe('community', () => {
    test('create', async () => {
      await create(req, res, next);
      let apiRes = res.json.mock.calls[0][0].toObject();

      newCommunityId = apiRes._id;
      apiRes = sanitize(apiRes, ['lastRewardFundUpdate', 'communityId']);
      const bobIsAdmin = apiRes.superAdmins.find(a => a.user.equals(bob._id));
      expect(bobIsAdmin).toBeDefined();
      expect(apiRes.image).toBe(brandNew.image);
      expect(apiRes).toMatchSnapshot();
    });

    test('join', async () => {
      req = {
        user: carol,
        params: brandNew
      };
      await join(req, res, next);
      let apiRes = res.json.mock.calls[0][0].toObject();

      apiRes = sanitize(apiRes, 'communityId');
      expect(apiRes).toMatchSnapshot();
    });

    test('should not allow update', async () => {
      req = {
        user: carol,
        body: { ...brandNew, superAdmins: ['carol'] }
      };
      const errFn = jest.fn();
      await update(req, res, errFn);
      expect(errFn).toHaveBeenCalled();
    });

    test('should update', async () => {
      const bobMember = await CommunityMember.findOne({
        communityId: newCommunityId,
        user: bob._id
      });
      req = {
        user: bob,
        body: { ...brandNew, superAdmins: ['carol'], _id: newCommunityId },
        communityMember: bobMember
      };
      await update(req, res, next);
      let apiRes = res.json.mock.calls[0][0].toObject();
      apiRes = sanitize(apiRes, ['communityId', 'lastRewardFundUpdate']);

      const carolIsAdmin = apiRes.superAdmins.find(a => a.user.equals(carol._id));
      expect(carolIsAdmin).toBeDefined();
      expect(apiRes).toMatchSnapshot();
    });

    test('leave', async () => {
      req = {
        user: carol,
        params: brandNew
      };
      await leave(req, res, next);
      expect(res.status).toHaveBeenCalledWith(200);
      const deleted = await CommunityMember.findOne({
        user: carol._id,
        communityId: brandNew._id
      });
      expect(deleted).toBe(null);
    });
  });
});
