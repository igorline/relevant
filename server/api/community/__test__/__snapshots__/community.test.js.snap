// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Communities community create 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "admins": Array [
    Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "community": "imnew",
      "communityId": "[SANITIZED]",
      "createdAt": "[SANITIZED]",
      "degree": 0,
      "deletedCommunity": false,
      "embeddedUser": Object {
        "_id": "[SANITIZED]",
        "handle": "alice",
        "image": "https://hips.hearstapps.com/elleuk.cdnds.net/15/37/2048x2730/2048x2730-e6-8278-458c32fb7c74-assets-elleuk-com-gallery-17063-taylor-swift-hairstyle-bob-getty-gallery-01-jpg.jpg",
        "name": "Alice",
      },
      "invites": 0,
      "pagerank": 0,
      "pagerankNeg": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "postDegree": 0,
      "relevanceRecord": Array [],
      "reputation": 0,
      "role": "admin",
      "superAdmin": true,
      "unread": 0,
      "updatedAt": "[SANITIZED]",
      "user": "a00000000000000000000000",
    },
  ],
  "avgTwitterScore": 0,
  "betEnabled": true,
  "channels": Array [],
  "createdAt": "[SANITIZED]",
  "currentPosts": 0,
  "currentShares": 0,
  "danglingConsumer": 0,
  "defaultPost": "link",
  "description": "Discussion of blockchain tech and ecosystem",
  "hidden": false,
  "id": "[SANITIZED]",
  "image": "",
  "lastRewardFundUpdate": "[SANITIZED]",
  "memberCount": 2,
  "name": "New Community",
  "negConsumer": 0,
  "postCount": 0,
  "private": false,
  "rewardFund": 0,
  "slug": "imnew",
  "superAdmins": Array [
    Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "community": "imnew",
      "communityId": "[SANITIZED]",
      "createdAt": "[SANITIZED]",
      "degree": 0,
      "deletedCommunity": false,
      "embeddedUser": Object {
        "_id": "[SANITIZED]",
        "handle": "alice",
        "image": "https://hips.hearstapps.com/elleuk.cdnds.net/15/37/2048x2730/2048x2730-e6-8278-458c32fb7c74-assets-elleuk-com-gallery-17063-taylor-swift-hairstyle-bob-getty-gallery-01-jpg.jpg",
        "name": "Alice",
      },
      "invites": 0,
      "pagerank": 0,
      "pagerankNeg": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "postDegree": 0,
      "relevanceRecord": Array [],
      "reputation": 0,
      "role": "admin",
      "superAdmin": true,
      "unread": 0,
      "updatedAt": "[SANITIZED]",
      "user": "a00000000000000000000000",
    },
    Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "community": "imnew",
      "communityId": "[SANITIZED]",
      "createdAt": "[SANITIZED]",
      "degree": 0,
      "deletedCommunity": false,
      "embeddedUser": Object {
        "_id": "[SANITIZED]",
        "handle": "bob",
        "image": "https://yt3.ggpht.com/a/AGF-l7_IPKJORqk1Wo7AR4jMhW15Ma-za_QDlRxMcQ=s900-c-k-c0xffffffff-no-rj-mo",
        "name": "Bob",
      },
      "invites": 0,
      "pagerank": 0,
      "pagerankNeg": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "postDegree": 0,
      "relevanceRecord": Array [],
      "reputation": 0,
      "role": "admin",
      "superAdmin": true,
      "unread": 0,
      "updatedAt": "[SANITIZED]",
      "user": "b00000000000000000000000",
    },
  ],
  "topPostShares": 0,
  "topics": Array [
    "co community topic",
  ],
  "twitterCount": 0,
  "updatedAt": "[SANITIZED]",
}
`;

exports[`Communities community join 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "community": "imnew",
  "communityId": "[SANITIZED]",
  "createdAt": "[SANITIZED]",
  "degree": 0,
  "deletedCommunity": false,
  "embeddedUser": Object {
    "_id": "[SANITIZED]",
    "handle": "carol",
    "image": "https://ichef.bbci.co.uk/news/640/cpsprodpb/A369/production/_87133814_87135704.jpg",
    "name": "Carol",
  },
  "invites": 0,
  "pagerank": 0,
  "pagerankNeg": 0,
  "pagerankRaw": 0,
  "pagerankRawNeg": 0,
  "postDegree": 0,
  "relevanceRecord": Array [],
  "reputation": 0,
  "role": "user",
  "superAdmin": false,
  "unread": 0,
  "updatedAt": "[SANITIZED]",
  "user": "c00000000000000000000000",
}
`;

exports[`Communities community should update 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "admins": Array [
    Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "community": "imnew",
      "communityId": "[SANITIZED]",
      "createdAt": "[SANITIZED]",
      "degree": 0,
      "deletedCommunity": false,
      "embeddedUser": Object {
        "_id": "[SANITIZED]",
        "handle": "alice",
        "image": "https://hips.hearstapps.com/elleuk.cdnds.net/15/37/2048x2730/2048x2730-e6-8278-458c32fb7c74-assets-elleuk-com-gallery-17063-taylor-swift-hairstyle-bob-getty-gallery-01-jpg.jpg",
        "name": "Alice",
      },
      "invites": 0,
      "pagerank": 0,
      "pagerankNeg": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "postDegree": 0,
      "relevanceRecord": Array [],
      "reputation": 0,
      "role": "admin",
      "superAdmin": false,
      "unread": 0,
      "updatedAt": "[SANITIZED]",
      "user": "a00000000000000000000000",
    },
  ],
  "avgTwitterScore": 0,
  "channels": Array [],
  "createdAt": "[SANITIZED]",
  "currentPosts": 0,
  "currentShares": 0,
  "danglingConsumer": 0,
  "description": "Discussion of blockchain tech and ecosystem",
  "id": "[SANITIZED]",
  "image": "",
  "lastRewardFundUpdate": "[SANITIZED]",
  "memberCount": 3,
  "name": "New Community",
  "negConsumer": 0,
  "postCount": 0,
  "rewardFund": 0,
  "slug": "imnew",
  "superAdmins": Array [
    Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "community": "imnew",
      "communityId": "[SANITIZED]",
      "createdAt": "[SANITIZED]",
      "degree": 0,
      "deletedCommunity": false,
      "embeddedUser": Object {
        "_id": "[SANITIZED]",
        "handle": "carol",
        "image": "https://ichef.bbci.co.uk/news/640/cpsprodpb/A369/production/_87133814_87135704.jpg",
        "name": "Carol",
      },
      "invites": 0,
      "pagerank": 0,
      "pagerankNeg": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "postDegree": 0,
      "relevanceRecord": Array [],
      "reputation": 0,
      "role": "user",
      "superAdmin": true,
      "unread": 0,
      "updatedAt": "[SANITIZED]",
      "user": "c00000000000000000000000",
    },
  ],
  "topPostShares": 0,
  "topics": Array [
    "co community topic",
  ],
  "twitterCount": 0,
  "updatedAt": "[SANITIZED]",
}
`;
